### ISSUE TITLE
[Proposal] Phase Space Definition

### ISSUE THREAD

**Issue by szabo137** [QEDjl-project/QEDcore.jl#50]
## The current state
`AbstractPhasespaceDefinition` is used more or less as a placeholder for more sophisticated concepts. Also `PhasespaceDefinition{AbstractCoordinateSystem, AbstractFrameOfReference}` is more a placeholder rather then being functional. 

## The dream state
In general, Phase space definitions should give structures to define a layout of the phase space without actually calculating all four-momenta . However, these layouts should provide a map of a set of coordinates to the momenta (also referred to as the momentum map). In the most general case, this should be done generically, i.e. orthogonal to process and model definitions. Eventually, the combination of process, model and phase space definition should fully determine the momentum map.
Additionally, it would be very convenient to have the possibility to define in-phase-space layouts independently from the out-phase-space layouts. This would also require to partially initialize the phase-space layout by fixing the incoming momenta. 

## Suggested solution 
From `AbstractPhaseSpaceDefiniton` as a root type, there are two kinds of phase space layouts: `AbstractInPhaseSpaceLayout` and `AbstractOutPhaseSpaceLayout{<: AbstractInPhaseSpaceLayout}`. As the type structure suggests, there is no out-phase-space layout without a corresponding in-layout. Subtypes of `AbstractInPhaseSpaceLayout` could be either just layouts (without initiation) or initialized with fixed in-momenta (maybe using `InPhaseSpacePoint`). This would make the corresponding out-phase-space layout "partially initialized". 
The actual momentum map could then be defined as an interface function on the respective layout: 

```Julia
_build_momenta(proc, model, in_ps_def, in_coords)
_build_momenta(proc, model, out_ps_def, in_coords, out_coords) # for non-initialized in_ps_def
_build_momenta(proc, model, out_ps_def, out_coords) # for initialized in_ps_def
```
These interface functions could then be used in the construction of `PhaseSpacePoint`.

## Final remarks 
I would suggest to exploit the name "layout" instead of "definition", because I think it is more reasonable and explicit. Furthermore, I suggest to "build" momenta instead of "generate" them.

The construction of the respective layout can exploit helper types like frame of reference, or coordinate systems. However, these should be done on specific branches in the type tree proposed above.

**Comment by szabo137** [comment]
@AntonReinhard what do out think?

**Comment by AntonReinhard** [comment]
It's probably a good idea to tackle this now :+1: 
I have a question though: If the `PhaseSpaceDefinition` will be used to construct the momenta, will it still be necessary to save it in the `PhaseSpacePoint` too? It seems to be only used in the  `_phase_space_factor` function, and maybe the result of that could just be cached in the PSP directly? Not sure if that would be better, but it would make the PSP type a bit flatter.

**Comment by szabo137** [comment]
I don't think it is a good idea to store the phase space factor in the psp without the actual layout, because it is a specific quantity, which has nothing to do with the psp in the first place. Furthermore, a reconstruction of the psp with different input is not possible, if the layout information is not stored. For example of you Lorentz boost a psp, having the layout would be very convenient, because one could move from a rest frame to a boosted frame in the layout as well.

**Comment by AntonReinhard** [comment]
Is this still relevant after https://github.com/QEDjl-project/QEDcore.jl/pull/94 or can this be closed now?

**Comment by szabo137** [comment]
Indeed, this can be closed.
